pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,104 @@
+package scheduler
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/the-maldridge/nbuild/pkg/graph"
+	"github.com/the-maldridge/nbuild/pkg/types"
+)
+
+func NewScheduler(l hclog.Logger, c CapacityProvider, url string) *Scheduler {
+	x := Scheduler{
+		l:                l.Named("scheduler"),
+		capacityProvider: c,
+		apiClient:        graph.NewAPIClient(l),
+		queue:            make([]Build, 0),
+		queueMutex:       new(sync.Mutex),
+		tuples:           make([]types.SpecTuple, 0),
+	}
+	x.apiClient.Url = url
+
+	return &x
+}
+
+// Pops a build off the queue and hands it off to the CapacityProvider.
+func (s *Scheduler) send() bool {
+	s.queueMutex.Lock()
+	defer s.queueMutex.Unlock()
+
+	if len(s.queue) == 0 {
+		return false
+	}
+	err := s.capacityProvider.DispatchBuild(s.queue[0])
+	if err != nil {
+		s.l.Trace("Unable to dispatch right now", "build", s.queue[0], "err", err)
+		return false
+	}
+	s.l.Trace("Dispatching", "build", s.queue[0])
+	s.queue = s.queue[1:]
+	return true
+}
+
+// Reconstructs the queue from dispatchable.
+func (s *Scheduler) Reconstruct() bool {
+	dispatchable, ok := s.apiClient.GetDispatchable()
+	if !ok {
+		return false
+	}
+
+	s.queueMutex.Lock()
+	defer s.queueMutex.Unlock()
+	s.queue = make([]Build, 0)
+
+	current, err := s.capacityProvider.ListBuilds()
+	if err != nil {
+		return false
+	}
+convert:
+	for tuple, pkgs := range dispatchable.Pkgs {
+		for _, pkg := range pkgs {
+			b := Build{
+				Spec: tuple,
+				Pkg:  pkg,
+				Rev:  dispatchable.Rev,
+			}
+			for _, curBuild := range current {
+				if b.Equal(&curBuild) {
+					break convert
```

Comment:
This is a code smell, lets try to find a way to refactor this so that it doesn't require the double level break.

Reply:
This is actually not even correct re-reading the code.